dot/sync/chain_sync.go

```diff
@@ -609,19 +678,41 @@ func (cs *chainSync) doSync(req *network.BlockRequestMessage) *workerError {
 	// response was validated! place into ready block queue
 	for _, bd := range resp.BlockData {
 		// block is ready to be processed!
-		logger.Trace("new ready block", "hash", bd.Hash, "number", bd.Header.Number)
-		cs.pendingBlocks.removeBlock(bd.Hash)
-		cs.readyBlocks <- bd
+		handleReadyBlock(bd, cs.pendingBlocks, cs.readyBlocks)
 	}
 
 	return nil
 }
 
+func handleReadyBlock(bd *types.BlockData, pendingBlocks DisjointBlockSet, readyBlocks *blockQueue) {
+	// see if there are any descendents in the pending queue that are now ready to be processed,
+	// as we have just become aware of their parent block
+
+	// if header was not requested, get it from the pending set
+	// if we're expecting headers, validate should ensure we have a header
+	if bd.Header == nil {
+		block := pendingBlocks.getBlock(bd.Hash)
+		if block.header == nil {
+			panic("block isn't ready! header is unknown!")
+		}
+
+		bd.Header = block.header
+	}
+
+	logger.Trace("new ready block", "hash", bd.Hash, "number", bd.Header.Number)
+
+	ready := []*types.BlockData{bd}
+	ready = pendingBlocks.getReadyDescendants(bd.Hash, ready)
+
+	for _, rb := range ready {
+		pendingBlocks.removeBlock(rb.Hash)
+		readyBlocks.push(rb)
+	}
+}
+
 // determineSyncPeers returns a list of peers that likely have the blocks in the given block request.
-// TODO: implement this
 func (cs *chainSync) determineSyncPeers(_ *network.BlockRequestMessage) []peer.ID {
```

Comment:
why do we need this `network.BlockRequestMessage` param?

Reply:
I need to update this so it checks the `peerState`s to check which peers have the blocks in the request. can do that in a follow-up